main/lsp/ErrorReporter.cc

```diff
@@ -0,0 +1,91 @@
+#include "main/lsp/ErrorReporter.h"
+#include "core/lsp/TypecheckEpochManager.h"
+#include "main/lsp/LSPMessage.h"
+#include "main/lsp/LSPOutput.h"
+#include "main/lsp/json_types.h"
+
+namespace sorbet::realmain::lsp {
+using namespace std;
+ErrorReporter::ErrorReporter(shared_ptr<const LSPConfiguration> config) : config(config) {}
+
+// Used for unit tests
+const std::vector<ErrorStatus> &ErrorReporter::getFileErrorStatuses() const {
+    return fileErrorStatuses;
+};
+
+void ErrorReporter::setMaxFileId(u4 id) {
+    ENFORCE(id >= fileErrorStatuses.size());
+    fileErrorStatuses.resize(id + 1, ErrorStatus{0, false});
+};
+
+void ErrorReporter::pushDiagnostics(u4 epoch, core::FileRef file, vector<unique_ptr<core::Error>> errors,
+                                    const core::GlobalState &gs) {
+    ENFORCE(file.exists());
+
+    if (file.id() >= fileErrorStatuses.size()) {
+        setMaxFileId(file.id());
+    }
+
+    ErrorStatus &fileErrorStatus = fileErrorStatuses[file.id()];
+
+    if (file.data(gs).epoch > epoch || fileErrorStatus.sentEpoch > epoch) {
+        return;
+    }
+    // If errors is empty and the file had no errors previously, break
+    if (errors.empty() && fileErrorStatus.hasErrors == false) {
+        return;
+    }
+
+    fileErrorStatus.hasErrors = !errors.empty();
+
+    if (fileErrorStatus.hasErrors) {
```

Comment:
There is no need for this conditional. `sentEpoch` is the last time we sent updated errors to VS Code for this file. We've already established that we're going to send diagnostics for this file (it'll just be the empty list).